extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -0,0 +1,612 @@
+        DEFAULT_FIRST_DAY_OF_WEEK;
+
+    this.daySize_ = this.element.getAttribute('day-size') || DEFAULT_DATE_SIZE;
+
+    const blocked = this.element.getAttribute('blocked');
+    /** @private @const */
+    this.blocked_ = new DatesList(
+        blocked ? blocked.split(DATE_SEPARATOR) : []);
+
+    const highlighted = this.element.getAttribute('highlighted');
+    /** @private @const */
+    this.highlighted_ = new DatesList(
+        highlighted ? highlighted.split(DATE_SEPARATOR) : []);
+
+    /** @private @const */
+    this.container_ = this.element.ownerDocument.createElement('div');
+
+    const type = this.element.getAttribute('type') || 'single';
+    /** @private @const */
+    this.picker_ = (type === 'range' ?
+        createDateRangePicker() :
+        createSingleDatePicker());
+
+    /** @private @const */
+    this.props_ = this.getProps_();
+
+    /** @private @const */
+    this.elementTemplates_ = this.parseElementTemplates_();
+
+    /** @private {!Array<!DateTemplateMapDef>} */
+    this.srcTemplates_ = [];
+
+    /** @private {?Element} */
+    this.srcDefaultTemplate_ = null;
+
+    /** @private @const */
+    this.isRTL_ = isRTL(/** @type {!Document} */ (this.ampdoc_.getRootNode()));
+
+    /** @private {?function(?):?} */
+    this.templateThen_ = null;
+
+    /** @private @const */
+    this.registerAction_ = this.registerAction.bind(this);
+
+    const locale = this.element.getAttribute('locale') || DEFAULT_LOCALE;
+    this.moment_.locale(locale);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    user().assert(isExperimentOn(this.win, TAG),
+        `Experiment ${TAG} is disabled.`);
+
+    this.element.appendChild(this.container_);
+    this.render();
+    this.element.setAttribute('i-amphtml-date-picker-attached', '');
+  }
+
+  /** @override */
+  layoutCallback() {
+    return this.parseSrcTemplates_(this.fetchSrcTemplates_());
+  }
+
+  /**
+   * Fetch the JSON from the URL specified in the src attribute.
+   * @return {?Promise<!JsonObject|!Array<JsonObject>>}
+   */
+  fetchSrcTemplates_() {
+    if (this.element.getAttribute('src')) {
+      return fetchBatchedJsonFor(this.ampdoc_, this.element);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Create an array of objects mapping dates to templates.
+   * @param {?Promise<!JsonObject|!Array<JsonObject>>} srcTemplatePromise
+   */
+  parseSrcTemplates_(srcTemplatePromise) {
+    if (!srcTemplatePromise) {
+      return Promise.resolve();
+    }
+
+    return srcTemplatePromise.then(srcJson => {
+      const templates = srcJson && srcJson['templates'];
+      if (!templates) {
+        return;
+      }
+
+      const srcTemplates = templates
+          .filter(t => t.dates)
+          .map(t => ({
+            dates: new DatesList(t.dates),
+            template: this.ampdoc_.getRootNode().querySelector(
```

Comment:
I assume templates can't be outside of `<amp-date-picker>` (mostly due to validator restrictions)., unless I am mistaken we should be able to use `this.element` for the query.

Reply:
I think `template-impl.js` supports templates outside the parent element via the `template=""` attribute, which queries for the `id` in the whole document, so I think it shouldn't be too problematic for the validator to do it elsewhere.

https://github.com/ampproject/amphtml/blob/4ab32dc304e33dacffc98ba3512a40aa5e39ebba/src/service/template-impl.js#L226-L233